File: multivendor-ecommerce/src/main/java/com/ecommerce/platform/service/CategoryService.java
package com.ecommerce.platform.service;

import com.ecommerce.platform.dto.CategoryRequest;
import com.ecommerce.platform.dto.CategoryResponse;
import com.ecommerce.platform.entity.Category;
import com.ecommerce.platform.exception.ResourceNotFoundException;
import com.ecommerce.platform.repository.CategoryRepository;
import jakarta.persistence.EntityExistsException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;

    @Transactional
    public CategoryResponse create(CategoryRequest request) {
        if (categoryRepository.existsByNameIgnoreCase(request.getName())) {
            throw new EntityExistsException("A category named '" + request.getName() + "' already exists");
        }
        Category category = Category.builder()
                .name(request.getName())
                .description(request.getDescription())
                .build();
        return toResponse(categoryRepository.save(category));
    }

    @Transactional
    public CategoryResponse update(Long id, CategoryRequest request) {
        Category category = getCategoryOrThrow(id);
        category.setName(request.getName());
        category.setDescription(request.getDescription());
        return toResponse(categoryRepository.save(category));
    }

    @Transactional
    public void delete(Long id) {
        Category category = getCategoryOrThrow(id);
        categoryRepository.delete(category);
    }

    public CategoryResponse getById(Long id) {
        return toResponse(getCategoryOrThrow(id));
    }

    public List<CategoryResponse> getAll() {
        return categoryRepository.findAll().stream().map(this::toResponse).toList();
    }

    private Category getCategoryOrThrow(Long id) {
        return categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found with id: " + id));
    }

    private CategoryResponse toResponse(Category category) {
        return CategoryResponse.builder()
                .id(category.getId())
                .name(category.getName())
                .description(category.getDescription())
                .build();
    }
}

File: multivendor_app/lib/services/category_service.dart
import 'package:dio/dio.dart';
import '../core/api_client.dart';
import '../models/category.dart';

class CategoryService {
  final Dio _dio = ApiClient.instance.dio;

  Future<List<Category>> getAll() async {
    try {
      final response = await _dio.get('/categories');
      return (response.data as List).map((e) => Category.fromJson(e)).toList();
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<Category> getById(int id) async {
    try {
      final response = await _dio.get('/categories/$id');
      return Category.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<Category> create({required String name, String? description}) async {
    try {
      final response = await _dio.post('/categories', data: {'name': name, 'description': description});
      return Category.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<Category> update(int id, {required String name, String? description}) async {
    try {
      final response =
          await _dio.put('/categories/$id', data: {'name': name, 'description': description});
      return Category.fromJson(response.data);
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }

  Future<void> delete(int id) async {
    try {
      await _dio.delete('/categories/$id');
    } catch (e) {
      throw ApiClient.instance.mapError(e);
    }
  }
}
